src/ir_def/ir.ml

```diff
@@ -95,7 +95,7 @@ and meta = {
   }
 
 and field = (field', Type.typ) Source.annotated_phrase
-and field' = {name : Type.lab; var : id} (* the var is by reference, not by value *)
+and field' = {name : Type.lab; var : id; byref : bool} (* the var is by reference, not by value *)
```

Comment:
Do you really need this field? The var was read by reference before already (to make DefineD work), so this ought to work fine for your aliasing with RefD just fine?

Reply:
The problem is that `is_mut` always used to allocate a fresh mutbox for a field. This flag makes it more like an immutable initialisation. I'd be super happy to get rid of it, but have no  better idea right now :-)

Reply:
Is it? The comment right there says something else, and I think in order to support await in record construction, at the IR level, all mutable fields are always initialized by reference, not by creating a new box.

I should look at the code on the laptop, though.

Reply:
I have an idea, will create a new issue for it… https://github.com/dfinity/motoko/issues/3351

Reply:
I didn't touch the comment. But I *think* it means "read the value off that variable", i.e. indirect. The construction of the outbox happens in `compile.ml:9160` where we forego the `Var.get_aliased_box` when that flag is set.

Reply:
Aaahh, if we could trick `get_aliased_box` into seeing a `HeapInd I` then it would do the trick for us, and `byref` would become redundant. I need to read up on ``HeapInd`!

Reply:
Exactly. And it’s not “tricking” – I think your `RefD` will invariably create something that is stored on the heap, as everything else would be wrong™.

So I think you don’t need `byref`, but “just” have to make sure that `AllocHow` does the right thing.

Reply:
Hmm, it’s fishy that `get_aliased_box` doesn’t just _always_ find a mutable box there. This relative dynamic alias-or-not is a bit dubious…

Reply:
I think there is a disconnect. `AllocHow.StoreHeap` (which makes `HeapInd`-accessed variables) is _not_ the correct notion here. Why? Because `HeapInd`-variables expect the trifecta of alloc/store-in/read-from of _mutboxes_, but what we want here is to have plain locals here that _deal with_ mutbox pointers! So `RefD` should default to `LocalImmut` (as it contains a plain pointer), but `get_aliased_box` should be prepared to deal with that by doing a `Var.get_val`.

Reply:
That catch all is bogus, see my other PR.

How do RefD and VarD differ? Don't both introduce a name of type Mut t that's represented by a pointer (in a local, of course) to a MutBox on the heap? That pointer is created differently (new box vs. reading a record), but afterwards I see no difference yet.

Reply:
I pushed my WIP as 19bbf3072. It contains vestiges of my `AllocHow.StoreHeap` experiment, which went wrong. Will cleanup later.

Reply:
Ok, but isn't that the same thing? 
The settable location from VarD is represented as a local with a pointer to a MutBox. I can read from the Box (VarE) or get the address of the box (NewObjE). Is that not enough?

Yes, it is a change of perspective: if you think of the RefD's _value_ as the VarD's _address_ they are different. But for the RefD you only care about it's address for putting it into an object, right? (or is this where I am confused?)

Reply:
Yes, `RefD` is just a pointer to a mutbox that wants to be stashed into a `var` field of a new record.

Reply:
Right, and so is a `VarD` :-)

Reply:
Yes, but `VarD` will create unique mutboxes, right? `RefD` avoids  that. Anyway, the effective changes in `compile.ml` are now down to the two additions about `RefD` and a few `compile_lexp` additions that will go away (when I change `RefD`'s `lexp` to `(lab, exp)`. The goodies are in 06674bae8. PTAL!